portfolio/src/main/java/com/google/sps/servlets/CommentsController.java

```diff
@@ -14,36 +14,239 @@
+
+    // Parse parameters
+    int page = parsePageParam(request.getParameter("page"));
+    int pageSize = parsePageSizeParam(request.getParameter("pageSize"));
+    SortDirection order = parseOrderParam(request.getParameter("order"));
+    String orderBy = parseOrderByParam(request.getParameter("orderBy"));
+
+    Query query = new Query("Comment").addSort(orderBy, order);
+    PreparedQuery results = datastore.prepare(query);
+
+    List<Comment> commentsList = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      long id = entity.getKey().getId();
+      String username = (String) entity.getProperty("username");
+      String comment = (String) entity.getProperty("comment");
+      long createdAt = (long) entity.getProperty("createdAt");
+      int likes = ((Long) entity.getProperty("likes")).intValue();
+
+      commentsList.add(new Comment(id, username, comment, createdAt, likes));
     }
+
+    int count = commentsList.size();
+    int pages = (int) Math.ceil((double) count / pageSize);
+    Integer next = page < pages ? (page + 1) : null;
+    Integer prev = page > 1 ? (page - 1) : null;
+
+    Info info = new Info(next, prev, count, pages);
+    int lowerIndex = (page - 1) * pageSize;
+    int upperIndex = page * pageSize;
+    List<Comment> trimmedComments = commentsList.subList(lowerIndex >= 0 ? lowerIndex : 0,
+        upperIndex <= count ? upperIndex : count);
+
+    Comments comments = new Comments(info, trimmedComments);
+    Gson gson = new Gson();
+    String data = gson.toJson(comments);
+
+    response.getWriter().println(data);
   }
 
   @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     response.setContentType("application/json;");
 
+    String body = getBody(request);
+    JSONObject json = new JSONObject();
+    try {
+      JSONParser parser = new JSONParser();
+      json = (JSONObject) parser.parse(body);
+    } catch (ParseException e) {
+      e.printStackTrace();
+    }
+
+    String username = getAttribute(json, "username", "Anonymous");
+    String comment = getAttribute(json, "comment", "...");
+    long createdAt = System.currentTimeMillis();
+    int likes = 0;
+
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("username", username);
+    commentEntity.setProperty("comment", comment);
+    commentEntity.setProperty("createdAt", createdAt);
+    commentEntity.setProperty("likes", likes);
+
+    datastore.put(commentEntity);
+
+    long id = commentEntity.getKey().getId();
+
     Gson gson = new Gson();
-    String data = gson.toJson(comments);
+    String data = gson.toJson(new Comment(id, username, comment, createdAt, likes));
+
+    response.getWriter().println(data);
+  }
+
+  @Override
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
```

Comment:
I added the votes functionality to the `PUT` request for the sake of consistency with the endpoint and its methods.

Although it only serves as a way to increment votes, I think that creating an endpoint in `comment/:commentId/vote` would be too much because a comment won't be edited in any other way.

If a refactor is needed, I can do that.

Reply:
This seems fine